{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Demultiplex Code\n",
    "\n",
    "Note: ran this on test files and then manually moved them into test_data_output\n",
    "\n",
    "Full files (non-test files) have a total of 363246735 reads per file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    ###### Create test files. Correct output of test files is ######\n",
    "              ####### only in regards to i1 and i2 #######\n",
    "#Total reads with N: 4            (forward/reverse reads 1 and 7)\n",
    "#Total low quality reads: 4      (forward/reverse reads 8, and 9)\n",
    "#Total index hopped reads: 4      (forward/reverse reads 4, 5)\n",
    "#Total correct reads: 6           (forward/reverse 3, 6, 10)\n",
    "#Total indexes w/out N and not matching 24 expected: 2      (forward/reverse reads 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/env python3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create variables to refer to each file as F (forward read), I1 (index 1), I2 (index2), R (reverse read)\n",
    "#python pathways start with /Users.. not /mnt/c/Users...\n",
    "\n",
    "# Method 1:    via arg parse\n",
    "# import argparse\n",
    "\n",
    "# def get_arguments():\n",
    "#     parser = argparse.ArgumentParser(description='demultiplex program for paired end reads and count instances of index hopping')\n",
    "#     parser.add_argument(\"-R1\", \"--read1_fq\", help =\"path to forward reads fastq\", required=True, type=str)\n",
    "#     parser.add_argument(\"-R2\", \"--read2_fq\", help =\"path to reverse reads fastq\", required=True, type=str)\n",
    "#     parser.add_argument(\"-I1\", \"--index1_fq\", help =\"path to forward index fastq\", required=True, type=str)\n",
    "#     parser.add_argument(\"-I2\", \"--index2_fq\", help =\"path to reverse index fastq\", required=True, type=str)\n",
    "#     parser.add_argument(\"-rds\", \"--total_reads\", help =\"total number of reads in a given input file. note this number needs to be the same across all input files\", required=True, type=int)\n",
    "#     return parser.parse_args()\n",
    "    \n",
    "# args = get_arguments()\n",
    "# read1 = args.read1_fq\n",
    "# read2 = args.read2_fq\n",
    "# index1 = args.index1_fq\n",
    "# index2 = args.index2_fq\n",
    "# num_reads = args.total_reads\n",
    "\n",
    "########################################## or use ##############################################################\n",
    "\n",
    "# Method 2:    via objects\n",
    "read1 = \"/Users/Jordan/UO_Docs/Genomics_lab/assignments/demultiplexing-jordan2lee/2_code/test_data/R1_testfile.fq\"\n",
    "read2 = \"/Users/Jordan/UO_Docs/Genomics_lab/assignments/demultiplexing-jordan2lee/2_code/test_data/R2_testfile.fq\"\n",
    "index1 = \"/Users/Jordan/UO_Docs/Genomics_lab/assignments/demultiplexing-jordan2lee/2_code/test_data/i1_testfile.fq\"\n",
    "index2 = \"/Users/Jordan/UO_Docs/Genomics_lab/assignments/demultiplexing-jordan2lee/2_code/test_data/i2_testfile.fq\"\n",
    "\n",
    "num_reads = 10"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create reverse compliment function\n",
    "def reverse_complement(barcode):\n",
    "    '''Takes string of seq and output string of reverse complementary'''\n",
    "    comp_bases = {'A':'T', 'C':'G', 'G':'C', 'T':'A', 'N':'N'}\n",
    "    return ''.join([comp_bases[base] for base in barcode[::-1]])\n",
    "\n",
    "#reverse_complement('CTAGCTCA') #test input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create function to convert scores to phred scores\n",
    "def convert_phred(letter):\n",
    "    \"\"\"Converts a single character into a phred score based on Illumina 1.8+Phred+33\"\"\"\n",
    "    ASCII = ord(letter) -33\n",
    "    return(ASCII)\n",
    "\n",
    "#convert_phred(\"J\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dictionary that has all 24 expected barcoes\n",
    "#index_dict: keys = barcodes   values= counts (start at 0)\n",
    "index_dict = {}\n",
    "b_init = [\"GTAGCGTA\", \"CGATCGAT\", \"GATCAAGG\", \"AACAGCGA\", \"TAGCCATG\", \"CGGTAATC\", \"CTCTGGAT\", \"TACCGGAT\", \\\n",
    "      \"CTAGCTCA\", \"CACTTCAC\", \"GCTACTCT\", \"ACGATCAG\", \"TATGGCAC\", \"TGTTCCGT\", \"GTCCTAAG\", \"TCGACAAG\", \\\n",
    "      \"TCTTCGAC\", \"ATCATGCG\", \"ATCGTGGT\", \"TCGAGAGT\", \"TCGGATTC\", \"GATCTTGC\", \"AGAGTCCA\", \"AGGATAGC\"]\n",
    "ct_init = [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]\n",
    "for i in range(len(b_init)):\n",
    "    index_dict[b_init[i]]= ct_init[i]\n",
    "#index_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create function to see if barcode in index_dict\n",
    "def check_barcode(index_seq):\n",
    "    '''Take string of index seq and see if in index_dict, returns boolean + input index_seq (True if index seq matches expected - returns False if index seq not expected)'''\n",
    "    for key in index_dict:\n",
    "        #print(key)\n",
    "        if index_seq == key:\n",
    "            #if input is a index in dict then return true + exit function\n",
    "            return index_seq == key\n",
    "        else:\n",
    "            #if don't find input seq in dict for that index of dict then pass \n",
    "            #and continue through function by repeating for next index in dict \n",
    "            pass\n",
    "    #returns False only when else statment is true -- because will exit function on \"if index_seq == key:\"     \n",
    "    return False\n",
    "                   \n",
    "check_barcode(\"AGGATAGC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    " # Create function to see if any bp has a qual score below qual score cutoff\n",
    "def qs_cutoff (quality_seq):\n",
    "    '''Take string of a full quality seq from fq/fa + convert to phred score, THEN return True if phred score >30 else false'''\n",
    "    pos = 0  # start with 1st qual in seq\n",
    "    #stop after reach end of quality seq\n",
    "    for pos in range(len(quality_seq)): \n",
    "        score = convert_phred(quality_seq[pos])\n",
    "        \n",
    "        #if phred score is = or < quality score cuttoff of 30f\n",
    "        if score <= 30:  \n",
    "            #return boolean False if any quality score is below or equal quality score cutoff\n",
    "            return(score>30)\n",
    "        \n",
    "        #if phred score is greater than quality score cutoff of 30\n",
    "        else:   \n",
    "            continue\n",
    "            pos+=1\n",
    "    \n",
    "    #return boolean True if all quality scores are above quality score cutoff\n",
    "    return(score>30)\n",
    "\n",
    "#qs_cutoff(\"AAAAJ#F\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Body of code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "LN = 0\n",
    "\n",
    "# Initialize counters to 0\n",
    "total_N = 0\n",
    "total_unexpected = 0\n",
    "total_low_qual = 0\n",
    "total_index_hop = 0\n",
    "total_correct = 0\n",
    "\n",
    "#create empty dictionary where key= index seq and values =file handle\n",
    "fh1_dict = {}\n",
    "fh2_dict = {}\n",
    "\n",
    "# OPEN all files\n",
    "with open(read1, \"r\") as r1, open(read2, \"r\") as r2, open(index1, \"r\") as i1, open(index2, \"r\") as i2, \\\n",
    "        open(\"undetermined_r1\", \"w\") as und_r1, open(\"undetermined_r2\", \"w\") as und_r2:\n",
    "    \n",
    "    # Opens file handles for that specific index seq\n",
    "    for index in b_init: \n",
    "        good_r1 = open(index + \"_r1.fq\", \"w\")\n",
    "        good_r2 = open(index + \"_r2.fq\", \"w\")\n",
    "        #create dictionary that has key=index seq and vlaues=file handle (dow this twice)\n",
    "        fh1_dict[index] = good_r1\n",
    "        fh2_dict[index] = good_r2\n",
    "\n",
    "    #tells program to read fastq files 4 lines at a time (header, seq, plus, qual) for ALL input files \n",
    "    while i1 and i2 and r1 and r2: # open while have  info \n",
    "        \n",
    "        # read i1 file 4 lines at a time\n",
    "        header_i1 = i1.readline().strip()\n",
    "        if not header_i1: #tells to exit loop when reach last line in file\n",
    "            break\n",
    "        seq_i1 = i1.readline().strip()\n",
    "        plus_i1 = i1.readline().strip()\n",
    "        qual_i1 = i1.readline().strip()\n",
    "        #lines_i1 = header_i1 + \"\\n\" + seq_i1  + \"\\n\" + plus_i1 + \"\\n\"  + qual_i1  #only for debugging\n",
    "        #print(lines_i1)\n",
    "\n",
    "        # read i2 file 4 lines at a time\n",
    "        header_i2 = i2.readline().strip()\n",
    "        seq_i2 = i2.readline().strip()\n",
    "        plus_i2 = i2.readline().strip()\n",
    "        qual_i2 = i2.readline().strip()\n",
    "        #lines_i2 = header_i2 + \"\\n\" + seq_i2  + \"\\n\" + plus_i2 + \"\\n\"  + qual_i2  #only for debugging\n",
    "        #print(lines_i2)\n",
    "\n",
    "        # read r1 file 4 lines at a time\n",
    "        header_r1 = r1.readline().strip()\n",
    "        seq_r1 = r1.readline().strip()\n",
    "        plus_r1 = r1.readline().strip()\n",
    "        qual_r1 = r1.readline().strip()\n",
    "        # Create lines obj so can call all 4 lines (aka all info of one read)\n",
    "        lines_r1 = header_r1 + \":\" + seq_i1 + \"\\n\" + seq_r1  + \"\\n\" + plus_r1 + \"\\n\"  + qual_r1\n",
    "        #print(lines_r1)\n",
    "        \n",
    "        # read r2 file 4 lines at a time\n",
    "        header_r2 = r2.readline().strip()\n",
    "        seq_r2 = r2.readline().strip()\n",
    "        plus_r2 = r2.readline().strip()\n",
    "        qual_r2 = r2.readline().strip()\n",
    "        lines_r2 = header_r2 + \":\" + seq_i2 + \"\\n\" + seq_r2  + \"\\n\" + plus_r2 + \"\\n\"  + qual_r2\n",
    "        #print(lines_r2)\n",
    "\n",
    "\n",
    "        # If either index seq has any N --> write to undeterminded file + add count to counter   \n",
    "        if \"N\" in seq_i1 or \"N\" in seq_i2:  # or reads as \"and/or\"\n",
    "            \n",
    "            # Test\n",
    "            #print(\"\\nReads that contain N\\n\", lines_i1, \"\\n\", lines_i2)\n",
    "            \n",
    "            # add all 4 lines of r1 read into undetermined_r1 file\n",
    "            und_r1.write(lines_r1 + \"\\n\") \n",
    "            total_N +=1  #add to counter for i1  \n",
    "            \n",
    "            # add all 4 lines of r2 read into undetermined_r2 file\n",
    "            und_r2.write(lines_r2 + \"\\n\")   \n",
    "            total_N+=1  #add to counter for i2 \n",
    "            \n",
    "                    \n",
    "        # if either index seq is not part of the 24 expected index seq --> write to undetermined file + add count to counter\n",
    "        elif check_barcode(seq_i1) == False or check_barcode(reverse_complement(seq_i2)) == False:\n",
    "\n",
    "            # Test\n",
    "            #print(\"\\n**Index unexpected AND new --> add to und_files**\\n\", lines_i1, \"\\n\", lines_i2)\n",
    "\n",
    "            # add all 4 lines of r1 read into undetermined_r1 file\n",
    "            und_r1.write(lines_r1 + \"\\n\")\n",
    "            total_unexpected +=1 # add to counter for i1\n",
    "\n",
    "            # add all 4 lines of r2 to undtermined_r2 file\n",
    "            und_r2.write(lines_r2 + \"\\n\")\n",
    "            total_unexpected +=1 # add to counter for i2\n",
    "                \n",
    "\n",
    "        # If any either index seq has a bp phred quality score is < quality score cutoff (30) -->  write to undetermined file + add to counter\n",
    "        elif qs_cutoff(qual_i1) == False or qs_cutoff(qual_i2) == False:\n",
    "\n",
    "            # add all 4 lines of r1 read into undetermined_r1 file\n",
    "            und_r1.write(lines_r1 + \"\\n\") \n",
    "            total_low_qual +=1  #add to counter for i1\n",
    "\n",
    "            # add all 4 lines of r2 read into undetermined_r2 file\n",
    "            und_r2.write(lines_r2 + \"\\n\") \n",
    "            total_low_qual+=1  #add to counter for i2\n",
    "                    \n",
    "            \n",
    "            \n",
    "        # If index 1 and 2 are not reverse complementary to each other --> write to undeter file + add count to counter\n",
    "        elif reverse_complement(seq_i2) != seq_i1 and \"N\" not in seq_i1 and \"N\" not in seq_i2:\n",
    "\n",
    "            # Test\n",
    "            #print(\"\\n**Index hop and new --> add to und_files**\\n\", lines_i1, \"\\n\", lines_i2)\n",
    "\n",
    "            # add all 4 lines of r1 read into undetermined_r1 file\n",
    "            und_r1.write(lines_r1 + \"\\n\") \n",
    "            total_index_hop +=1  #add to counter for i1\n",
    "\n",
    "            # add all 4 lines of r2 read into undetermined_r2 file\n",
    "            und_r2.write(lines_r2 + \"\\n\") \n",
    "            total_index_hop+=1  #add to counter for i2\n",
    "\n",
    "\n",
    "        # If index 1 and 2 are reverse complementary to each other --> write to \"good\" file\n",
    "        elif reverse_complement(seq_i2) == seq_i1:\n",
    "\n",
    "            # test\n",
    "            #print(\"\\n**Reads that are good**\\n\", lines_i1, \"\\n\", lines_i2)\n",
    "\n",
    "            #create file with <barcode>_r1.fq\n",
    "            #good_r1 = open(seq_i1 + \"_r1.fq\", \"a\") \n",
    "            # add all 4 lines of r1 read into undetermined_r1 file\n",
    "            \n",
    "            #grab value from dict -- this is the fh\n",
    "            good_r1 = fh1_dict[seq_i1]\n",
    "            good_r1.write(lines_r1 + \"\\n\") \n",
    "            total_correct +=1  #add to counter for i1\n",
    "\n",
    "            #create file with <reverse complement of barcode>_r2.fq\n",
    "            #so that matches with the forward read barcode (e.g. good_r1)\n",
    "            #good_r2 = open(reverse_complement(seq_i2) + \"_r2.fq\", \"a\") \n",
    "            # add all 4 lines of r2 read into undetermined_r2 file\n",
    "            good_r2 = fh2_dict[reverse_complement(seq_i2)]\n",
    "            good_r2.write(lines_r2 + \"\\n\") \n",
    "            total_correct+=1  #add to counter for i2 \n",
    "        \n",
    "    #close file handles for that specific index seq\n",
    "    for index in b_init:\n",
    "        fh1_dict[index].close()\n",
    "        fh2_dict[index].close()\n",
    "                    \n",
    "    #TODO close all files\n",
    "    \n",
    "    #Create a useful report for the end user of your code\n",
    "    str_N =str(total_N)\n",
    "    str_N_prop = str(total_N/num_reads)\n",
    "    str_low_qual = str(total_low_qual)\n",
    "    str_low_qual_prop = str(total_low_qual/num_reads)\n",
    "    str_hop = str(total_index_hop)\n",
    "    str_hop_prop = str(total_index_hop/num_reads)\n",
    "    str_good = str(total_correct)\n",
    "    str_good_prop = str(total_correct/num_reads)\n",
    "    str_une = str(total_unexpected)\n",
    "    str_une_prop = str(total_unexpected/num_reads)\n",
    "    \n",
    "    with open(\"report_demultiplex\", \"w\") as report:\n",
    "        report.write(\"Total reads passed (good): \" + str_good + \"\\n\" + \"Proportion reads passed (good): \" + str_good_prop + \"\\n\")\n",
    "        report.write(\"Total reads with N: \" + str_N + \"\\n\" + \"Proportion reads with N: \" + str_N_prop + \"\\n\")\n",
    "        report.write(\"Total reads with low quality: \" + str_low_qual + \"\\n\" + \"Proportion reads with low quality: \" + str_low_qual_prop + \"\\n\")\n",
    "        report.write(\"Total reads with index hopping: \" + str_hop + \"\\n\" + \"Proportion reads with index hopping: \" + str_hop_prop + \"\\n\")\n",
    "        report.write(\"Total reads without N and not matching expected barcodes: \" + str_une + \"\\n\" + \"Proportion reads without N and not matching expected barcodes: \" + str_une_prop)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Unit Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Unit Tests #######\n",
    "\"\"\"Check that stand alone functions are working properly\"\"\"\n",
    "assert reverse_complement('CACTTCAC') == 'GTGAAGTG'\n",
    "assert reverse_complement('TGTTCCGT') == 'ACGGAACA'\n",
    "assert reverse_complement('CTAGCTCA') != 'AGAGCTAA'\n",
    "\n",
    "assert convert_phred(\"/\") == 14\n",
    "assert convert_phred(\"#\") == 2\n",
    "assert convert_phred(\"G\") != 31\n",
    "\n",
    "assert \"CTAGCTCA\" in index_dict\n",
    "assert \"TCTTCGAC\" in index_dict\n",
    "assert \"AAAAAAAA\" not in index_dict\n",
    "\n",
    "assert check_barcode(\"AGGATAGC\") == True\n",
    "assert check_barcode(\"TATGGCAC\") == True\n",
    "assert check_barcode(\"TATGACAA\") == False\n",
    "\n",
    "assert qs_cutoff(\"AAAAJKF\") == True\n",
    "assert qs_cutoff(\"A/AAJBF\") == False\n",
    "assert qs_cutoff(\"AAAAJ#F\") == False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Unit Test: Specific to Exact Test Files Created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "##### Unit Test: correct program should output this file summary  #######\n",
    "          ####### only in regards to i1 and i2 #######\n",
    "#Total reads with N: 4            (forward/reverse reads 1 and 7)\n",
    "#Total low quality reads: 4      (forward/reverse reads 8, and 9)\n",
    "#Total index hopped reads: 4      (forward/reverse reads 4, 5)\n",
    "#Total correct reads: 6           (forward/reverse 3, 6, 10)\n",
    "#Total indexes w/out N and not matching 24 expected: 2      (forward/reverse reads 2)\n",
    "\n",
    "####### Test file specific Unit Test: only works with my specific test files #########W\n",
    "#\"\"\"Check that total counts of each type are correct\"\"\"\n",
    "assert total_N == 4\n",
    "assert total_low_qual == 4\n",
    "assert total_index_hop == 4\n",
    "assert total_correct == 6\n",
    "assert total_unexpected == 2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
